README.md
<p align="center">
  <h1 align="center">cypress-web-vitals</h1>
</p>
<p align="center">
A <a href="https://github.com/GoogleChrome/web-vitals">Web Vitals</a> command for <a href="https://github.com/cypress-io/cypress">Cypress</a>.
</p>
<p align="center">
   <a href="https://github.com/cmorten/cypress-web-vitals/tags/"><img src="https://img.shields.io/github/tag/cmorten/cypress-web-vitals" alt="cypress-web-vitals versions" /></a>
   <a href="https://www.npmjs.com/package/cypress-web-vitals"><img alt="cypress-web-vitals available on NPM" src="https://img.shields.io/npm/dy/cypress-web-vitals"></a>
   <img src="https://github.com/cmorten/cypress-web-vitals/workflows/Test/badge.svg" alt="Current test status" />
   <a href="http://makeapullrequest.com"><img src="https://img.shields.io/badge/PRs-welcome-brightgreen.svg" alt="PRs are welcome" /></a>
   <a href="https://github.com/cmorten/cypress-web-vitals/issues/"><img src="https://img.shields.io/github/issues/cmorten/cypress-web-vitals" alt="cypress-web-vitals issues" /></a>
   <img src="https://img.shields.io/github/stars/cmorten/cypress-web-vitals" alt="cypress-web-vitals stars" />
   <img src="https://img.shields.io/github/forks/cmorten/cypress-web-vitals" alt="cypress-web-vitals forks" />
   <img src="https://img.shields.io/github/license/cmorten/cypress-web-vitals" alt="cypress-web-vitals license" />
   <a href="https://GitHub.com/cmorten/cypress-web-vitals/graphs/commit-activity"><img src="https://img.shields.io/badge/Maintained%3F-yes-green.svg" alt="cypress-web-vitals is maintained" /></a>
</p>
<p align="center">
  <i>Quantifying the quality of user experience on your site.</i>
</p>

---

## About

[Web Vitals](https://web.dev/vitals/) is a Google initiative which provides a set of quality signals and thresholds that are essential to delivering a great user experience on the web.

`cypress-web-vitals` allows you to test against these signals within your Cypress workflows through a set of [custom commands](https://docs.cypress.io/api/cypress-api/custom-commands):

- `cy.vitals()` - self contained command for performing a Web Vitals audit of page load performance.
- `cy.startVitalsCapture()` - command for starting a journey based Web Vitals audit, enabling capture of [interaction to next paint (INP)](https://web.dev/inp/).
- `cy.reportVitals()` - command for reporting on a journey based Web Vitals audit started with `cy.startVitalsCapture()`.

## Getting started

### Install the dependencies

In your terminal:

```console
$ yarn add -D cypress-web-vitals cypress-real-events
# or
$ npm install --save-dev cypress-web-vitals cypress-real-events
```

_Note: `cypress-web-vitals` currently makes use of `cypress-real-events` to click the page as a real user would to calculate [first input delay](https://web.dev/fid/). Hence it is needed as a peer-dependency._

### Add the commands

Add the following line to your `cypress/support/commands.js`:

```js
import "cypress-web-vitals";
```

### Write your tests

```js
describe("Web Vitals", () => {
  it("should pass the audits for a page load", () => {
    cy.vitals({ url: "https://www.google.com/" });
  });

  it("should pass the audits for a customer journey", () => {
    cy.startVitalsCapture({
      url: "https://www.google.com/",
    });

    cy.findByRole("combobox", { name: "Search" }).realClick();
    cy.findByRole("listbox").should("be.visible");

    cy.reportVitals();
  });
});
```

_Note: this example is making use of the [Cypress Testing Library](https://testing-library.com/docs/cypress-testing-library/intro) for interacting with the page. This is not required for use of this package nor is it included in the package._

## Examples

Example Cypress test setups with a variety of tests using `cypress-web-vitals` for Cypress 9.x, 10.x, and 12.x are available in the [`./examples` directory](./examples).

## API

### `cy.vitals([WebVitalsConfig])`

Performs and audit against the Google Web Vitals.

```js
cy.vitals();
cy.vitals(webVitalsConfig);
```

Example:

```js
cy.vitals({ firstInputSelector: "main" }); // Use the `main` element of the page for clicking to capture the FID.
cy.vitals({ thresholds: { cls: 0.2 } }); // Test the page against against a CLS threshold of 0.2.
```

`WebVitalsConfig`:

- `Optional` **firstInputSelector**: `string | string[]` - Selector(s) for the element(s) to click for capturing FID. Can be a single element selector or an array, all of which will be clicked. For each selector the first matching element is used. Default: `"body"`.
- `Optional` **onReport**: `Function` - Callback for custom handling of the report results, e.g. for sending results to application monitoring platforms.
- `Optional` **strict**: `boolean` - Enables strict mode in which tests will fail if metrics with specified thresholds cannot be calculated.
- `Optional` **thresholds**: `WebVitalsThresholds` - The thresholds to audit the Web Vitals against. If not provided Google's 'Good' scores will be used (see below). If provided, any missing Web Vitals signals will not be audited.
- `Optional` **url**: `string` - The url to audit. If not provided you will need to have called `cy.visit(url)` prior to the command.
- `Optional` **headers**: `string` - Additional headers that will be provided to `cy.visit()` if `url` is provided.
- `Optional` **auth**: `string` - Additional auth that will be provided to `cy.visit()` if `url` is provided.
- `Optional` **vitalsReportedTimeout**: `number` - Time in ms to wait for Web Vitals to be reported before failing. Default: `10000`.

### `cy.startVitalsCapture([StartWebVitalsCaptureConfig])`

Starts an audit against the Google Web Vitals.

```js
cy.startVitalsCapture();
cy.startVitalsCapture(startWebVitalsCaptureConfig);
```

Example:

```js
cy.startVitalsCapture({
  url: "https://www.google.com/",
});
```

`StartWebVitalsCaptureConfig`:

- `Optional` **url**: `string` - The url to audit. If not provided you will need to have called `cy.visit(url)` prior to the command.
- `Optional` **headers**: `string` - Additional headers that will be provided to `cy.visit()` if `url` is provided.
- `Optional` **auth**: `string` - Additional auth that will be provided to `cy.visit()` if `url` is provided.

### `cy.reportVitals([ReportWebVitalsConfig])`

Completes and reports on an audit against the Google Web Vitals.

```js
cy.reportVitals();
cy.reportVitals(reportWebVitalsConfig);
```

```js
cy.reportVitals({ thresholds: { inp: 500 } }); // Test the page against against an INP threshold of 500.
```

`ReportWebVitalsConfig`:

- `Optional` **onReport**: `Function` - Callback for custom handling of the report results, e.g. for sending results to application monitoring platforms.
- `Optional` **strict**: `boolean` - Enables strict mode in which tests will fail if metrics with specified thresholds cannot be calculated.
- `Optional` **thresholds**: `WebVitalsThresholds` - The thresholds to audit the Web Vitals against. If not provided Google's 'Good' scores will be used (see below). If provided, any missing Web Vitals signals will not be audited.
- `Optional` **vitalsReportedTimeout**: `number` - Time in ms to wait for Web Vitals to be reported before failing. Default: `10000`.

### `WebVitalsThresholds`

- `Optional` **lcp**: `number` - Threshold for [largest contentful paint (LCP)](https://web.dev/lcp/).
- `Optional` **fid**: `number` - Threshold for [first input delay (FID)](https://web.dev/fid/).
- `Optional` **cls**: `number` - Threshold for [cumulative layout shift (CLS)](https://web.dev/cls/).
- `Optional` **fcp**: `number` - Threshold for [first contentful paint (FCP)](https://web.dev/fcp/).
- `Optional` **ttfb**: `number` - Threshold for [time to first byte (TTFB)](https://web.dev/time-to-first-byte/).
- `Optional` **inp**: `number` - Threshold for [interaction to next paint (INP)](https://web.dev/inp/).

#### Google 'Good' scores

```json
{
  "lcp": 2500,
  "fid": 100,
  "cls": 0.1,
  "fcp": 1800,
  "ttfb": 600,
  "inp": 200
}
```

### Custom Reporting

It can be useful to have direct access to the raw data so you can generate custom reports, send to APM, etc.

This can be achieved through the optional `onReport` callback for `cy.vitals()` or `cy.reportVitals()` which receives the raw report before `cypress-web-vitals` passes or fails the test.

```js
describe("Web Vitals", () => {
  it("should log the report to APM", () => {
    cy.vitals({
      url: "https://www.google.com/",
      onReport({ results, strict, thresholds }) {
        console.log(results); // { lcp: ..., fid: ..., }
      },
    });
  });
});
```

The report results contains values for _all signals_, not just the values specified in the provided or default `thresholds`. Signals that couldn't be obtained are reported as `null`.

## How does it work?

When using the `cy.vitals()` command:

1. The url is visited with the HTML response intercepted and modified by Cypress to include the [Web Vitals](https://github.com/GoogleChrome/web-vitals#from-a-cdn) module script and some code to record the Web Vitals values.
1. The audit then waits for the page load event to allow for the values of LCP and CLS to settle; which are subject to change as different parts of the page load into view.
1. Several elements (if exist) starting with the provided element(s) (based on `firstInputSelector`) are then clicked in quick succession to simulate a user clicking on the page to aid FID reporting. Note: if choosing a custom element(s), don't pick something that will navigate away from the page otherwise the plugin will fail to capture the Web Vitals metrics.
1. Next the audit simulates a page visibility state change [which is required for the CLS web-vital to be reported](https://www.npmjs.com/package/web-vitals#basic-usage).
1. The audit then attempts to wait for any outstanding Web Vitals to be reported for which thresholds have been provided.
1. Finally the Web Vitals values are compared against the thresholds, logging successful results and throwing an error for any unsuccessful signals. Note: if the audit was unable to record a web-vital then it is logged, _but the test will not fail_.

The `cy.startVitalsCapture()` and `cy.reportVitals()` commands perform steps 1-2 and 4-6 respectively. There is no clicking of elements (step 3) with these commands as the expectation is for you to provide your own customer journey interactions.

## Contributing

Please check out the [CONTRIBUTING](./docs/CONTRIBUTING.md) docs.

## Changelog

Please check out the [CHANGELOG](./docs/CHANGELOG.md) docs.

---

## License

`cypress-web-vitals` is licensed under the [MIT License](./LICENSE.md).


docs/CHANGELOG.md
# ChangeLog

For the latest changes, please refer to the [releases page](https://github.com/cmorten/cypress-web-vitals/releases).

## [0.4.0] - 02-11-2021

- [#1] bug: fix script injection when head tag has whitespace or attributes (#2) @jonthemiller
- feat: `vitalsReportedTimeout` config (#3)

## [0.3.0] - 26-06-2021

- fix: `visibilityState` changes now simulated through stubbing the document as `window.open()` does not work with headless Cypress.
- docs: added section to readme on how the command works internally.
- feat: now wait for page load to allow LCP and CLS to change and settle.
- feat: now attempt to wait for all desired Web Vitals to have been reported before calculating the results.

## [0.2.0] - 26-06-2021

- ci: fix `.npmignore`.

## [0.1.0] - 26-06-2021

- feat: initial release.


src/constants.js
const DEFAULT_THRESHOLDS = {
  lcp: 2500,
  fid: 100,
  cls: 0.1,
  fcp: 1800,
  ttfb: 600,
  inp: 200,
};

const WEB_VITALS_KEYS = Object.keys(DEFAULT_THRESHOLDS);
const WEB_VITALS_KEYS_WITHOUT_CLS_FID_LCP_INP = WEB_VITALS_KEYS.filter(
  (vital) => !["cls", "fid", "lcp", "inp"].includes(vital),
);

const DEFAULT_FIRST_INPUT_SELECTOR = "body";

const DEFAULT_STRICT_MODE = false;

const SUPPORTED_BROWSERS = {
  Chrome: true,
  Chromium: true,
  Canary: true,
};

const WEB_VITALS_ACCESSOR_KEY = "__cy_web_vitals__";

// https://unpkg.com/web-vitals@3.4.0/dist/web-vitals.iife.js
const WEB_VITALS_CORE = `var webVitals=function(e){"use strict";var n,t,r,i,o,a=-1,c=function(e){addEventListener("pageshow",(function(n){n.persisted&&(a=n.timeStamp,e(n))}),!0)},u=function(){return window.performance&&performance.getEntriesByType&&performance.getEntriesByType("navigation")[0]},s=function(){var e=u();return e&&e.activationStart||0},f=function(e,n){var t=u(),r="navigate";a>=0?r="back-forward-cache":t&&(document.prerendering||s()>0?r="prerender":document.wasDiscarded?r="restore":t.type&&(r=t.type.replace(/_/g,"-")));return{name:e,value:void 0===n?-1:n,rating:"good",delta:0,entries:[],id:"v3-".concat(Date.now(),"-").concat(Math.floor(8999999999999*Math.random())+1e12),navigationType:r}},d=function(e,n,t){try{if(PerformanceObserver.supportedEntryTypes.includes(e)){var r=new PerformanceObserver((function(e){Promise.resolve().then((function(){n(e.getEntries())}))}));return r.observe(Object.assign({type:e,buffered:!0},t||{})),r}}catch(e){}},l=function(e,n,t,r){var i,o;return function(a){n.value>=0&&(a||r)&&((o=n.value-(i||0))||void 0===i)&&(i=n.value,n.delta=o,n.rating=function(e,n){return e>n[1]?"poor":e>n[0]?"needs-improvement":"good"}(n.value,t),e(n))}},p=function(e){requestAnimationFrame((function(){return requestAnimationFrame((function(){return e()}))}))},v=function(e){var n=function(n){"pagehide"!==n.type&&"hidden"!==document.visibilityState||e(n)};addEventListener("visibilitychange",n,!0),addEventListener("pagehide",n,!0)},m=function(e){var n=!1;return function(t){n||(e(t),n=!0)}},h=-1,g=function(){return"hidden"!==document.visibilityState||document.prerendering?1/0:0},T=function(e){"hidden"===document.visibilityState&&h>-1&&(h="visibilitychange"===e.type?e.timeStamp:0,C())},y=function(){addEventListener("visibilitychange",T,!0),addEventListener("prerenderingchange",T,!0)},C=function(){removeEventListener("visibilitychange",T,!0),removeEventListener("prerenderingchange",T,!0)},E=function(){return h<0&&(h=g(),y(),c((function(){setTimeout((function(){h=g(),y()}),0)}))),{get firstHiddenTime(){return h}}},L=function(e){document.prerendering?addEventListener("prerenderingchange",(function(){return e()}),!0):e()},b=[1800,3e3],S=function(e,n){n=n||{},L((function(){var t,r=E(),i=f("FCP"),o=d("paint",(function(e){e.forEach((function(e){"first-contentful-paint"===e.name&&(o.disconnect(),e.startTime<r.firstHiddenTime&&(i.value=Math.max(e.startTime-s(),0),i.entries.push(e),t(!0)))}))}));o&&(t=l(e,i,b,n.reportAllChanges),c((function(r){i=f("FCP"),t=l(e,i,b,n.reportAllChanges),p((function(){i.value=performance.now()-r.timeStamp,t(!0)}))})))}))},w=[.1,.25],P=function(e,n){n=n||{},S(m((function(){var t,r=f("CLS",0),i=0,o=[],a=function(e){e.forEach((function(e){if(!e.hadRecentInput){var n=o[0],t=o[o.length-1];i&&e.startTime-t.startTime<1e3&&e.startTime-n.startTime<5e3?(i+=e.value,o.push(e)):(i=e.value,o=[e])}})),i>r.value&&(r.value=i,r.entries=o,t())},u=d("layout-shift",a);u&&(t=l(e,r,w,n.reportAllChanges),v((function(){a(u.takeRecords()),t(!0)})),c((function(){i=0,r=f("CLS",0),t=l(e,r,w,n.reportAllChanges),p((function(){return t()}))})),setTimeout(t,0))})))},F={passive:!0,capture:!0},I=new Date,A=function(e,i){n||(n=i,t=e,r=new Date,k(removeEventListener),M())},M=function(){if(t>=0&&t<r-I){var e={entryType:"first-input",name:n.type,target:n.target,cancelable:n.cancelable,startTime:n.timeStamp,processingStart:n.timeStamp+t};i.forEach((function(n){n(e)})),i=[]}},D=function(e){if(e.cancelable){var n=(e.timeStamp>1e12?new Date:performance.now())-e.timeStamp;"pointerdown"==e.type?function(e,n){var t=function(){A(e,n),i()},r=function(){i()},i=function(){removeEventListener("pointerup",t,F),removeEventListener("pointercancel",r,F)};addEventListener("pointerup",t,F),addEventListener("pointercancel",r,F)}(n,e):A(n,e)}},k=function(e){["mousedown","keydown","touchstart","pointerdown"].forEach((function(n){return e(n,D,F)}))},B=[100,300],x=function(e,r){r=r||{},L((function(){var o,a=E(),u=f("FID"),s=function(e){e.startTime<a.firstHiddenTime&&(u.value=e.processingStart-e.startTime,u.entries.push(e),o(!0))},p=function(e){e.forEach(s)},h=d("first-input",p);o=l(e,u,B,r.reportAllChanges),h&&v(m((function(){p(h.takeRecords()),h.disconnect()}))),h&&c((function(){var a;u=f("FID"),o=l(e,u,B,r.reportAllChanges),i=[],t=-1,n=null,k(addEventListener),a=s,i.push(a),M()}))}))},N=0,R=1/0,H=0,O=function(e){e.forEach((function(e){e.interactionId&&(R=Math.min(R,e.interactionId),H=Math.max(H,e.interactionId),N=H?(H-R)/7+1:0)}))},_=function(){return o?N:performance.interactionCount||0},j=function(){"interactionCount"in performance||o||(o=d("event",O,{type:"event",buffered:!0,durationThreshold:0}))},q=[200,500],V=0,z=function(){return _()-V},G=[],J={},K=function(e){var n=G[G.length-1],t=J[e.interactionId];if(t||G.length<10||e.duration>n.latency){if(t)t.entries.push(e),t.latency=Math.max(t.latency,e.duration);else{var r={id:e.interactionId,latency:e.duration,entries:[e]};J[r.id]=r,G.push(r)}G.sort((function(e,n){return n.latency-e.latency})),G.splice(10).forEach((function(e){delete J[e.id]}))}},Q=function(e,n){n=n||{},L((function(){j();var t,r=f("INP"),i=function(e){e.forEach((function(e){(e.interactionId&&K(e),"first-input"===e.entryType)&&(!G.some((function(n){return n.entries.some((function(n){return e.duration===n.duration&&e.startTime===n.startTime}))}))&&K(e))}));var n,i=(n=Math.min(G.length-1,Math.floor(z()/50)),G[n]);i&&i.latency!==r.value&&(r.value=i.latency,r.entries=i.entries,t())},o=d("event",i,{durationThreshold:n.durationThreshold||40});t=l(e,r,q,n.reportAllChanges),o&&(o.observe({type:"first-input",buffered:!0}),v((function(){i(o.takeRecords()),r.value<0&&z()>0&&(r.value=0,r.entries=[]),t(!0)})),c((function(){G=[],V=_(),r=f("INP"),t=l(e,r,q,n.reportAllChanges)})))}))},U=[2500,4e3],W={},X=function(e,n){n=n||{},L((function(){var t,r=E(),i=f("LCP"),o=function(e){var n=e[e.length-1];n&&n.startTime<r.firstHiddenTime&&(i.value=Math.max(n.startTime-s(),0),i.entries=[n],t())},a=d("largest-contentful-paint",o);if(a){t=l(e,i,U,n.reportAllChanges);var u=m((function(){W[i.id]||(o(a.takeRecords()),a.disconnect(),W[i.id]=!0,t(!0))}));["keydown","click"].forEach((function(e){addEventListener(e,u,!0)})),v(u),c((function(r){i=f("LCP"),t=l(e,i,U,n.reportAllChanges),p((function(){i.value=performance.now()-r.timeStamp,W[i.id]=!0,t(!0)}))}))}}))},Y=[800,1800],Z=function e(n){document.prerendering?L((function(){return e(n)})):"complete"!==document.readyState?addEventListener("load",(function(){return e(n)}),!0):setTimeout(n,0)},$=function(e,n){n=n||{};var t=f("TTFB"),r=l(e,t,Y,n.reportAllChanges);Z((function(){var i=u();if(i){var o=i.responseStart;if(o<=0||o>performance.now())return;t.value=Math.max(o-s(),0),t.entries=[i],r(!0),c((function(){t=f("TTFB",0),(r=l(e,t,Y,n.reportAllChanges))(!0)}))}}))};return e.CLSThresholds=w,e.FCPThresholds=b,e.FIDThresholds=B,e.INPThresholds=q,e.LCPThresholds=U,e.TTFBThresholds=Y,e.getCLS=P,e.getFCP=S,e.getFID=x,e.getINP=Q,e.getLCP=X,e.getTTFB=$,e.onCLS=P,e.onFCP=S,e.onFID=x,e.onINP=Q,e.onLCP=X,e.onTTFB=$,Object.defineProperty(e,"__esModule",{value:!0}),e}({});`;

const WEB_VITALS_REPORTING = `
(w => {
  w.${WEB_VITALS_ACCESSOR_KEY} = w.${WEB_VITALS_ACCESSOR_KEY} || {};
  const e = new Event("${WEB_VITALS_ACCESSOR_KEY}");
  webVitals.onCLS(m => { console.log("CLS", m); w.${WEB_VITALS_ACCESSOR_KEY}.cls=m; w.dispatchEvent(e) });
  webVitals.onFID(m => { console.log("FID", m); w.${WEB_VITALS_ACCESSOR_KEY}.fid=m; w.dispatchEvent(e) });
  webVitals.onLCP(m => { console.log("LCP", m); w.${WEB_VITALS_ACCESSOR_KEY}.lcp=m; w.dispatchEvent(e) });
  webVitals.onFCP(m => { console.log("FCP", m); w.${WEB_VITALS_ACCESSOR_KEY}.fcp=m; w.dispatchEvent(e) });
  webVitals.onTTFB(m => { console.log("TTFB", m); w.${WEB_VITALS_ACCESSOR_KEY}.ttfb=m; w.dispatchEvent(e) });
  webVitals.onINP(m => { console.log("INP", m); w.${WEB_VITALS_ACCESSOR_KEY}.inp=m; w.dispatchEvent(e) });
})(window)
`;

const WEB_VITALS_SNIPPET = `<script>${WEB_VITALS_CORE}${WEB_VITALS_REPORTING}</script>`;

const LOG_SLUG = "cy.vitals()";

const DEFAULT_ALL_WEB_VITALS_REPORTED_TIMEOUT_MS = 10000;

module.exports = {
  DEFAULT_ALL_WEB_VITALS_REPORTED_TIMEOUT_MS,
  DEFAULT_FIRST_INPUT_SELECTOR,
  DEFAULT_STRICT_MODE,
  DEFAULT_THRESHOLDS,
  LOG_SLUG,
  SUPPORTED_BROWSERS,
  WEB_VITALS_ACCESSOR_KEY,
  WEB_VITALS_KEYS,
  WEB_VITALS_KEYS_WITHOUT_CLS_FID_LCP_INP,
  WEB_VITALS_SNIPPET,
};
